<!-- This comment will put IE 6, 7 and 8 in quirks mode -->
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<title>libmgf: libmgf: Mascot Generic Format (MGF) Parser</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javaScript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css"/>
</head>
<body onload='searchBox.OnSelectItem(0);'>
<!-- Generated by Doxygen 1.6.1 -->
<script type="text/javascript"><!--
var searchBox = new SearchBox("searchBox", "search",false,'Search');
--></script>
<div class="navigation" id="top">
  <div class="tabs">
    <ul>
      <li class="current"><a href="index.html"><span>Main&nbsp;Page</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li><a href="files.html"><span>Files</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <img id="MSearchSelect" src="search/search.png"
             onmouseover="return searchBox.OnSearchSelectShow()"
             onmouseout="return searchBox.OnSearchSelectHide()"
             alt=""/>
        <input type="text" id="MSearchField" value="Search" accesskey="S"
             onfocus="searchBox.OnSearchFieldFocus(true)" 
             onblur="searchBox.OnSearchFieldFocus(false)" 
             onkeyup="searchBox.OnSearchFieldChange(event)"/>
        <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
        </div>
      </li>
    </ul>
  </div>
</div>
<div class="contents">
<h1>libmgf: Mascot Generic Format (MGF) Parser</h1><h2><a class="anchor" id="sec_intro">
Introduction</a></h2>
<p><code>libmgf</code> (formerly mgfp) is a flex/bison-based C++ MGF parser library.<br/>
 It includes the library code as well as the following set of MGF processing tools:</p>
<ul>
<li>mgfvalidate -- validates a given MGF file </li>
<li>rix -- reporter ion extractor for iTRAQ4 and TMT </li>
<li>ms2preproc -- MGF preprocessing tool (see below)</li>
</ul>
<p>The <a href="http://steenlab.org">Steen &amp; Steen Lab</a> provides the library under the terms of a MIT license for use in academic and non-academic environments.</p>
<h2><a class="anchor" id="sec_citation">
Citation</a></h2>
<p>If you make use of <code>libmgf</code> in your own projects, please cite the following article:<br/>
</p>
<ul>
<li>Kirchner M, Steen JAJ, Hamprecht FA, Steen H (2010). MGFp: An open Mascot Generic Format parser library implementation. <em>Journal of Proteome Research, 9(5):2762-2763 (2010)</em>. [<a href="http://www.ncbi.nlm.nih.gov/pubmed/20334363">PubMed entry</a>]</li>
</ul>
<p>If you use <code>ms2preproc</code> in your data analysis pipeline, please cite<br/>
</p>
<ul>
<li>Renard BY, Kirchner M, Monigatti F, Invanov AR, Rappsilber J, Winter D, Steen JAJ, Hamprecht FA, Steen H, When Less Can Yield More - Computational Preprocessing of MS/MS Spectra for Peptide Identification, <em>Proteomics 9(21):4978-4984 (2009)</em>. [<a href="http://www.ncbi.nlm.nih.gov/pubmed/19743429">PubMed entry</a>]</li>
</ul>
<h2><a class="anchor" id="sec_install">
Installation</a></h2>
<h3><a class="anchor" id="sec_install_bin">
Obtaining the Software</a></h3>
<p>Binary packages for Microsoft Windows are available here: </p>
<ul>
<li><a href="https://github.com/kirchnerlab/libmgf/downloads">https://github.com/kirchnerlab/libmgf/downloads</a></li>
</ul>
<p>Linux and Mac users, please build from source.</p>
<h3><a class="anchor" id="sec_install_src">
Building from Source</a></h3>
<p>Building <code>libmgf</code> from source is straightforward. It requires a working CMake build system (available from <a href="http://cmake.org/">http://cmake.org/</a>) and CMake &gt;= 2.6.</p>
<p>With cmake available, the build process is </p>
<div class="fragment"><pre class="fragment">
 git clone git://github.com/kirchnerlab/libmgf.git
 mkdir libmgf-build
 cd libmgf-build
 cmake ../libmgf
 make
 make test
 make install
 make package (optional, generates binary packages for your platform)
</pre></div><p>For Mac users: if you are using MacPorts, then linking errors with <code>boost::program_options</code> seem common; alas, this is not a <code>libmgf</code> issue but indicates a compiler/library incompatibility on your system.</p>
<h2><a class="anchor" id="Usage">
Usage</a></h2>
<h3><a class="anchor" id="cmdline">
Using the command line applications</a></h3>
<p>All command line applications come with a <code>--help</code> switch that describes their usage.</p>
<h3><a class="anchor" id="lib">
Using the libmgf library</a></h3>
<p>To use the parser, one must first create a parser driver instance: </p>
<div class="fragment"><pre class="fragment"><span class="preprocessor">    #include &lt;mgf/mgf.h&gt;</span>
    ...
    mgf::MgfFile mgfFile;
    mgf::Driver driver(mgfFile);
</pre></div><p> Then set the verbosity flage (defaulting to off/false) </p>
<div class="fragment"><pre class="fragment">    driver.trace_parsing = <span class="keyword">true</span>;
    driver.trace_scanning = <span class="keyword">true</span>;
</pre></div><p> and parse the input. The input is a stream (like <code>std::cin</code> in the example here or <code>std::fstream</code>). </p>
<div class="fragment"><pre class="fragment">    <span class="keywordtype">bool</span> result = driver.parse_stream(std::cin);
</pre></div><p> One should always check if the parsing was successful and only continue if so. </p>
<div class="fragment"><pre class="fragment">    <span class="keywordflow">if</span> (!result) {
        std::cerr &lt;&lt; std::endl
          &lt;&lt; <span class="stringliteral">&quot;Error parsing data stream (use -v for details).&quot;</span> &lt;&lt; std::endl;
        <span class="keywordflow">return</span> -1;
    }
</pre></div><p> If the parsing was successful, the contents of the MGF file are available in terms of an MgfFile object; it is possible to iterate over the MS/MS spectra and to read/modify/otherwise process the contents. The example here attempts to extract TMT reporter ion intensities from centroid mode MS/MS spectra:</p>
<div class="fragment"><pre class="fragment">    <span class="keyword">typedef</span> mgf::MgfFile::iterator MFI;
    <span class="keywordflow">for</span> (MFI i = mgfFile.begin(); i != mgfFile.end(); ++i) {
        <span class="comment">// sort the MS/MS spectrum by m/z</span>
        std::sort(i-&gt;begin(), i-&gt;end(), mgf::LessThanMz());
        <span class="keyword">typedef</span> mgf::MgfSpectrum::iterator MSI;
        <span class="comment">// extract TMT reporter ion intensities</span>
        std::tr1::array&lt;double, 6&gt; obsTmtAbundances;
        <span class="keywordflow">for</span> (<span class="keywordtype">size_t</span> n = 0; n &lt; 6; ++n) {
            MSI closestIt = findClosestMz(i-&gt;begin(), i-&gt;end(), tmtMasses[n]);
            <span class="comment">// check if the closest centroid is close enough</span>
            <span class="keywordflow">if</span> (std::abs(closestIt-&gt;first - tmtMasses[n]) &lt; 0.5) {
                obsTmtAbundances[n] = closestIt-&gt;second;
            } <span class="keywordflow">else</span> {
                obsTmtAbundances[n] = 0.0;
            }
            tmts.push_back(obsTmtAbundances);
        }
    }
</pre></div><h2><a class="anchor" id="sec_examples">
Examples</a></h2>
<p>Coding examples are in the <code>applications/</code> subdirectory</p>
<h2><a class="anchor" id="sec_appendix">
Appendix</a></h2>
<h3><a class="anchor" id="sec_issues">
Known Issues</a></h3>
<ul>
<li>the MGF format has been misused many times; although MatrixScience provides a format description, there is no official formal grammar. As such, libmgf may sometimes refuse to accept a specific MGF dialect. Please contact the developers if that seems to be the case for you and we will see what we can do. Note that libmgf provides comprehensive error messages; please include this information in bug reports or enhancement requests. </li>
<li>Multiple USERxx entries; the current implementation ignores USER entries. </li>
<li>General PMF support: most libmgf testing has been done on MS/MS ion search data. Hence, it is possible that valid PMF constructs are not recognized. If that is the case, please contact the developers and we will attempt to fix issues as quickly as possible.</li>
</ul>
<h3><a class="anchor" id="sec_grammar">
The MGF Grammar</a></h3>
<p>The following is the current MGF Grammar, extracted from Parser.ypp. </p>
<div class="fragment"><pre class="fragment">0 $accept: start "end of file"

1 ion: "double" "double" "end of line"
2    | "integer" "double" "end of line"
3    | "double" "integer" "end of line"
4    | "integer" "integer" "end of line"

5 ions: ions ion
6     | ion

7 charge: "integer" '+'
8       | "integer" '-'

9 charges: '(' charges ')'
10        | charges ',' charge
11        | charges "and keyword" charge
12        | charge

13 csintegerlist: csintegerlist ',' "integer"
14              | "integer"

15 blocks: /* empty */
16       | blocks block

17 block: "begin_ions keyword" "end of line" localparams ions "end_ions keyword" "end of line"
18      | "begin_ions keyword" "end of line" localparams "end_ions keyword" "end of line"

19 globalparams: /* empty */
20             | globalparams globalparam

21 globalparam: "enzyme keyword" '=' "string" "end of line"
22            | "search title keyword" '=' "string" "end of line"
23            | "database keyword" '=' "string" "end of line"
24            | "MS/MS datafile format keyword" '=' "string" "end of line"
25            | "MS/MS ion series keyword" '=' "string" "end of line"
26            | "variable modifications keyword" '=' "string" "end of line"
27            | "units for ITOL keyword" '=' "string" "end of line"
28            | "mass type (mono or avg) keyword" '=' "string" "end of line"
29            | "fixed modifications keyword" '=' "string" "end of line"
30            | "quantitation method keyword" '=' "string" "end of line"
31            | "maximum hits keyword" '=' "string" "end of line"
32            | "type of report keyword" '=' "string" "end of line"
33            | "type of search keyword" '=' "string" "end of line"
34            | "taxonomy keyword" '=' "string" "end of line"
35            | "tolerance units keyword" '=' "string" "end of line"
36            | "user keyword" '=' "string" "end of line"
37            | "user email keyword" '=' "string" "end of line"
38            | "username keyword" '=' "string" "end of line"
39            | "perform decoy search keyword" '=' "integer" "end of line"
40            | "error tolerance keyword" '=' "integer" "end of line"
41            | "partials keyword" '=' "integer" "end of line"
42            | "fragment ion tolerance keyword" '=' "double" "end of line"
43            | "fragment ion tolerance keyword" '=' "integer" "end of line"
44            | "misassigned 13C keyword" '=' "double" "end of line"
45            | "precursor m/z keyword" '=' "double" "end of line"
46            | "precursor m/z keyword" '=' "integer" "end of line"
47            | "protein mass (kDa) keyword" '=' "double" "end of line"
48            | "protein mass (kDa) keyword" '=' "integer" "end of line"
49            | "peptide mass tolerance keyword" '=' "double" "end of line"
50            | "peptide mass tolerance keyword" '=' "integer" "end of line"
51            | "charge set keyword" '=' charges "end of line"
52            | "NA translation keyword" '=' csintegerlist "end of line"
53            | "comment" "end of line"

54 localparams: /* empty */
55            | localparams localparam

56 localparam: "title keyword and full title string" "end of line"
57           | "amino acid composition keyword" '=' "string" "end of line"
58           | "MS/MS ion series keyword" '=' "string" "end of line"
59           | "variable modifications keyword" '=' "string" "end of line"
60           | "retention time or range keyword" '=' "double" "end of line"
61           | "retention time or range keyword" '=' "integer" "end of line"
62           | "retention time or range keyword" '=' "double" '-' "double" "end of line"
63           | "retention time or range keyword" '=' "double" '-' "integer" "end of line"
64           | "retention time or range keyword" '=' "integer" '-' "double" "end of line"
65           | "retention time or range keyword" '=' "integer" '-' "integer" "end of line"
66           | "scan number of range keyword" '=' "integer" "end of line"
67           | "scan number of range keyword" '=' "integer" '-' "integer" "end of line"
68           | "tolerance units keyword" '=' "string" "end of line"
69           | "amino acid sequence keyword" '=' "string" "end of line"
70           | "sequence tag keyword" '=' "string" "end of line"
71           | "error tolerant sequence keyword" '=' "string" "end of line"
72           | "peptide mass tolerance keyword" '=' "double" "end of line"
73           | "peptide mass tolerance keyword" '=' "integer" "end of line"
74           | "charge set keyword" '=' charges "end of line"
75           | "precursor mass keyword" '=' "double" "end of line"
76           | "precursor mass keyword" '=' "double" "double" "end of line"
77           | "precursor mass keyword" '=' "double" "integer" "end of line"
78           | "precursor mass keyword" '=' "integer" "end of line"
79           | "precursor mass keyword" '=' "integer" "double" "end of line"
80           | "precursor mass keyword" '=' "integer" "integer" "end of line"
81           | "comment" "end of line"

82 contents: globalparams blocks "end of file"

83 start: contents


</pre></div> </div>
<!--- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&nbsp;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&nbsp;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&nbsp;</span>Functions</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&nbsp;</span>Variables</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&nbsp;</span>Typedefs</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<hr size="1"/><address style="text-align: right;"><small>Generated on Thu Nov 24 07:44:09 2011 for libmgf by&nbsp;
<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.6.1 </small></address>
</body>
</html>
